src/connectivity/reachability.c

```diff
@@ -0,0 +1,247 @@
+ *    components will be stored here.
+ * \param reach A list of bitsets representing the result. It will be resized
+ *    as needed. <code>reach[membership[u]][v]</code> is set to \c true if
+ *    vertex \c v is reachable from vertex \c u.
+ * \param directed For a directed graph, determines whether edges are treated
+ *    as directed or undirected. Ignored for undirected graphs.
+ * \return Error code:
+ *         \c IGRAPH_ENOMEM if there is not enough memory
+ *         to perform the operation.
+ *
+ * \sa \ref igraph_connected_components(), \ref igraph_count_reachable().
+ *
+ * Time complexity: O(|C||V|/w + |V| + |E|), where
+ * |C| is the number of strongly connected components (at most |V|),
+ * |V| is the number of vertices, and
+ * |E| is the number of edges respectively,
+ * and w is the bit width of \type igraph_integer_t, typically the
+ * word size of the machine (32 or 64).
+ */
+
+igraph_error_t igraph_reachability(
+        const igraph_t *graph,
+        igraph_vector_int_t *membership,
+        igraph_vector_int_t *csize,
+        igraph_integer_t *no_of_components,
+        igraph_bitset_list_t *reach,
+        igraph_bool_t directed) {
+
+    const igraph_integer_t no_of_nodes = igraph_vcount(graph);
+    igraph_integer_t no_of_comps;
+    igraph_adjlist_t adjlist, dag;
+
+    if (! igraph_is_directed(graph)) {
+        directed = false;
+    }
+
+    IGRAPH_CHECK(igraph_connected_components(graph,
+                                             membership, csize, &no_of_comps,
+                                             directed ? IGRAPH_STRONG : IGRAPH_WEAK));
+
+    if (no_of_components) {
+        *no_of_components = no_of_comps;
+    }
+
+    IGRAPH_CHECK(igraph_bitset_list_resize(reach, no_of_comps));
+
+    for (igraph_integer_t comp = 0; comp < no_of_comps; comp++) {
+        IGRAPH_CHECK(igraph_bitset_resize(igraph_bitset_list_get_ptr(reach, comp), no_of_nodes));
+    }
+    for (igraph_integer_t v = 0; v < no_of_nodes; v++) {
+        IGRAPH_BIT_SET(*igraph_bitset_list_get_ptr(reach, VECTOR(*membership)[v]), v);
+    }
+
+    if (!directed) {
+        return IGRAPH_SUCCESS;
+    }
+
+    IGRAPH_CHECK(igraph_adjlist_init(graph, &adjlist, IGRAPH_OUT, IGRAPH_LOOPS_ONCE, IGRAPH_MULTIPLE));
+    IGRAPH_FINALLY(igraph_adjlist_destroy, &adjlist);
+
+    IGRAPH_CHECK(igraph_adjlist_init_empty(&dag, no_of_comps));
+    IGRAPH_FINALLY(igraph_adjlist_destroy, &dag);
+
+    for (igraph_integer_t v = 0; v < no_of_nodes; v++) {
+        const igraph_vector_int_t *neighbours = igraph_adjlist_get(&adjlist, v);
+        igraph_vector_int_t *dag_neighbours = igraph_adjlist_get(&dag, VECTOR(*membership)[v]);
+        const igraph_integer_t n = igraph_vector_int_size(neighbours);
+        for (igraph_integer_t i = 0; i < n; i++) {
+            if (VECTOR(*membership)[v] != VECTOR(*membership)[VECTOR(*neighbours)[i]]) {
+                igraph_vector_int_push_back(dag_neighbours, VECTOR(*membership)[VECTOR(*neighbours)[i]]);
```

Comment:
Aren't we missing an `IGRAPH_CHECK()` here?

Reply:
@Tagl Could you please fix this? All function that return an error code should have an `IGRAPH_CHECK`, _unless_ we know for certain that they cannot fail in this specific use case.

Reply:
```suggestion
                IGRAPH_CHECK(igraph_vector_int_push_back(dag_neighbours, VECTOR(*membership)[VECTOR(*neighbours)[i]]));
```